File: src/data/o1/Median_of_Two_Sorted_Arrays/Median_of_Two_Sorted_Arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        // Ensure nums1 is the smaller array
        if (nums1.length > nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }
        
        int m = nums1.length;
        int n = nums2.length;
        int total = m + n;
        int half = (total + 1) / 2;
        
        int left = 0;
        int right = m;
        
        while (left <= right) {
            int partitionX = (left + right) / 2;
            int partitionY = half - partitionX;
            
            int maxLeftX = (partitionX == 0) ? Integer.MIN_VALUE : nums1[partitionX - 1];
            int minRightX = (partitionX == m) ? Integer.MAX_VALUE : nums1[partitionX];
            
            int maxLeftY = (partitionY == 0) ? Integer.MIN_VALUE : nums2[partitionY - 1];
            int minRightY = (partitionY == n) ? Integer.MAX_VALUE : nums2[partitionY];
            
            if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
                if (total % 2 == 0) {
                    return ((double)Math.max(maxLeftX, maxLeftY) + Math.min(minRightX, minRightY)) / 2;
                } else {
                    return (double)Math.max(maxLeftX, maxLeftY);
                }
            } else if (maxLeftX > minRightY) {
                right = partitionX - 1;
            } else {
                left = partitionX + 1;
            }
        }
        
        throw new IllegalArgumentException("Input arrays are not sorted.");
    }
}

public class Median_of_Two_Sorted_Arrays {
// Method to run all test cases
    public static void runTests(Solution solution) {
        // Test case 1: Both arrays are of the same size and have a clear median
        int[] nums1 = {1, 3};
        int[] nums2 = {2};
        double expected_result = 2.0;
        assert solution.findMedianSortedArrays(nums1, nums2) == expected_result : "Test case 1 failed. Expected " + expected_result;

        // Test case 2: Arrays of different sizes with an even combined length
        nums1 = new int[]{1, 2};
        nums2 = new int[]{3, 4};
        expected_result = 2.5;
        assert solution.findMedianSortedArrays(nums1, nums2) == expected_result : "Test case 2 failed. Expected " + expected_result;

        // Test case 3: One array is empty
        nums1 = new int[]{};
        nums2 = new int[]{1};
        expected_result = 1.0;
        assert solution.findMedianSortedArrays(nums1, nums2) == expected_result : "Test case 3 failed. Expected " + expected_result;

        // Test case 4: Both arrays have the same elements
        nums1 = new int[]{1, 1};
        nums2 = new int[]{1, 1};
        expected_result = 1.0;
        assert solution.findMedianSortedArrays(nums1, nums2) == expected_result : "Test case 4 failed. Expected " + expected_result;

        // Test case 5: Arrays with negative and positive numbers
        nums1 = new int[]{-5, 3, 6};
        nums2 = new int[]{-2, 0, 9};
        expected_result = 1.5;
        assert solution.findMedianSortedArrays(nums1, nums2) == expected_result : "Test case 5 failed. Expected " + expected_result;

        // Test case 6: Large arrays to check efficiency
        nums1 = new int[100000];
        nums2 = new int[100000];
        for (int i = 0; i < 100000; i++) {
            nums1[i] = i;
            nums2[i] = 100000 + i;
        }
        expected_result = 99999.5;
        assert solution.findMedianSortedArrays(nums1, nums2) == expected_result : "Test case 6 failed. Expected " + expected_result;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 200000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    

File: src/data/o1/Median_of_Two_Sorted_Arrays/Median_of_Two_Sorted_Arrays.cpp
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        // Ensure nums1 is the smaller array
        if (nums1.size() > nums2.size())
            return findMedianSortedArrays(nums2, nums1);
        
        int m = nums1.size();
        int n = nums2.size();
        int total = m + n;
        int half = (total + 1) / 2;
        
        int left = 0, right = m;
        while (left <= right){
            int i = left + (right - left) / 2;
            int j = half - i;
            
            int nums1Left = (i == 0) ? INT_MIN : nums1[i-1];
            int nums1Right = (i == m) ? INT_MAX : nums1[i];
            int nums2Left = (j == 0) ? INT_MIN : nums2[j-1];
            int nums2Right = (j == n) ? INT_MAX : nums2[j];
            
            if (nums1Left <= nums2Right && nums2Left <= nums1Right){
                if (total % 2 == 0){
                    return (max(nums1Left, nums2Left) + min(nums1Right, nums2Right)) / 2.0;
                }
                else{
                    return max(nums1Left, nums2Left);
                }
            }
            else if (nums1Left > nums2Right){
                right = i -1;
            }
            else{
                left = i +1;
            }
        }
        
        return 0.0;
    }
};

//  Test cases for the C++ solution
#include <cassert>

void testSolution(Solution& solution) {
    // Test case 1: Both arrays are of the same size and have a clear median
    std::vector<int> nums1 = {1, 3};
    std::vector<int> nums2 = {2};
    double expected_result = 2.0;
    assert(solution.findMedianSortedArrays(nums1, nums2) == expected_result && "Test case 1 failed.");

    // Test case 2: Arrays of different sizes with an even combined length
    nums1 = {1, 2};
    nums2 = {3, 4};
    expected_result = 2.5;
    assert(solution.findMedianSortedArrays(nums1, nums2) == expected_result && "Test case 2 failed.");

    // Test case 3: One array is empty
    nums1 = {};
    nums2 = {1};
    expected_result = 1.0;
    assert(solution.findMedianSortedArrays(nums1, nums2) == expected_result && "Test case 3 failed.");

    // Test case 4: Both arrays have the same elements
    nums1 = {1, 1};
    nums2 = {1, 1};
    expected_result = 1.0;
    assert(solution.findMedianSortedArrays(nums1, nums2) == expected_result && "Test case 4 failed.");

    // Test case 5: Arrays with negative and positive numbers
    nums1 = {-5, 3, 6};
    nums2 = {-2, 0, 9};
    expected_result = 1.5;
    assert(solution.findMedianSortedArrays(nums1, nums2) == expected_result && "Test case 5 failed.");

    // Test case 6: Large arrays to check efficiency
    nums1.clear();
    nums2.clear();
    for (int i = 0; i < 100000; ++i) nums1.push_back(i);
    for (int i = 100000; i < 200000; ++i) nums2.push_back(i);
    expected_result = 99999.5;
    assert(solution.findMedianSortedArrays(nums1, nums2) == expected_result && "Test case 6 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 1280; ++i) {
        testSolution(solution);  
    }
    return 0;
}
